商品購入機能

55Suape opened the pull request:
#What
商品購入機能の実装

#Why
商品購入機能の実装を行いましたので、ご確認をお願いします。

- [ ] 必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/646dd4084402ac0d6d74c0d6d9961ec5

- [ ] 入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/c14d60715cc7b4bfa0a7847c349e06e0

- [ ] ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/b3d881f2628ac4c40a281ff92e77abc2

- [ ] ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/890d61276bd79bb0e7d9db0dc723ba74

- [ ] ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/3f542876a2565b4327f65fceb1fbe881

- [ ] 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/17cf4124b3f7366e0bc754e278a0c404

- [ ] 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/f0f880b4bf54aea4e6097b9d581b3677

- [ ] ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/c709620d3fb4538d8c69a492de965482

- [ ] ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/5d0cefc7626a7e68b7bbe674187f8868

- [ ] テスト結果の画像
https://gyazo.com/58d29b40b9092fb512e2058adc02b949

55Suape committed:
Formオブジェクトを作成

app/modelsディレクトリ直下にorder_address.rbファイルを作成

55Suape committed:
コントローラーとモデルの作成＋Viewとルーティングの修正

55Suape committed:
単体テスト

55Suape committed:
PAY.JPの導入とOrder_addressモデルの単体テスト

55Suape committed:
売却有無での表示切り替えとアクセス制限

55Suape committed:
rubocop

OhyaYouki commented on spec/factories/order_addresses.rb:
```diff
@@ -0,0 +1,12 @@
+FactoryBot.define do
+  factory :order_address do
+    user_id       { Faker::Number.non_zero_digit }
+    item_id       { Faker::Number.non_zero_digit }
```
```ruby
    user_id       { Faker::Number.non_zero_digit }
    item_id       { Faker::Number.non_zero_digit }
```
※
user_idやitem_idといった外部キーに数値を直接渡すのはあまり良い実装ではありません。
まずはユーザーや商品の情報を作成し、そのidをuser_idとitem_idに渡してあげるのがより良い実装となります。

 今回の購入機能であれば、order_delivery_spec.rbのbeforeブロック内でuserやitemの情報を作成してあげましょう。
以下のような記述をすることが考えられます。

```rb
before do
  user = FactoryBot.create(:user)
  @order_delivery = FactoryBot.build(:order_delivery, user_id: user.id)
  ##今回の例ではuserしか紐付けを行っていません。itemとも紐付けを行いましょう。
end
```

参考カリキュラム：
最終課題補足カリキュラム
〜1つのフォームから複数のテーブルに情報を保存するアプリを作成しよう〜


また、この修正によってテスト実行時に`mysql connection error`のようなものが表示される場合は、`config/environments/test.rb`のファイルを修正する必要があります。

参考カリキュラム：
応用カリキュラム
Ruby on Rails（ChatApp）
〜ChatAppのモデル単体テストコードを書こう〜


ご不明な点がありましたら、コネクトにてメンターにご相談ください。
よろしくお願い致します。

OhyaYouki commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,71 @@
+        @order_address.token = nil
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Token can't be blank")
+      end
```
以下のテストを追加しましょう。
```ruby
・userが紐づいていないと購入できない
・itemが紐づいていないと購入できない
```

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。

コメントいたしましたので、ご確認をお願いいたします。

コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
ご対応のほど、よろしくお願いいたします。

55Suape committed:
単体テストの追記とmysql connection errorエラー改修

##spec/factories/order_addresses.rb
- 外部キーに直接数値を渡す記述トルツメ

##spec/models/order_address_spec.rb
- beforeブロック内のFactoryBotにuser_id(user情報),item_id(item情報)を渡す指示を追記
- userが紐付いていないと購入できない項目追記
- itemが紐付いていないと購入できない項目追記

##config/environments/test.rb
- mysql connection errorエラー対策

55Suape commented:
# What
商品購入機能の実装

# Why
コードレビューありがとうございます。
各種修正を反映したため、ご確認をお願いします。

## spec/factories/order_addresses.rb
- 外部キーに直接値を渡す記述トルツメ

## spec/models/order_address_spec.rb
- beforeブロック内のFactoryBotにuser_id(user情報),item_id(item情報)を渡す指示を追記
- userが紐付いていないと購入できない項目追記
- itemが紐付いていないと購入できない項目追記

## config/environments/test.rb
- mysql connection errorエラー対策

Kentozzz commented on app/controllers/items_controller.rb:
```diff
@@ -22,9 +22,9 @@ def create
+  # 出品ユーザー = カレントユーザー且つSold Outになっていない場合は編集画面へ遷移、それ以外はトップページへ遷移
   def edit
-    if @item.user == current_user
+    if @item.user == current_user && @item.order.nil?
```
```
if @item.user == current_user && @item.order.nil?
```
上記の記述は一部`destroy`アクションと同じ条件式を使用しているためprivate以下でメソッド化して記述をまとめることができます。
まとめた後はbefore_actionで使用しましょう！

Kentozzz commented on app/models/order_address.rb:
```diff
@@ -0,0 +1,25 @@
+    validates :city, format: { with: /\A[ぁ-んァ-ン一-龥]/, message: 'is invalid. Input full-width characters.' }
+    validates :house_number
+    VALID_PHONE_REGEX = /\A\d{10}$|^\d{11}\z/
+    validates :phone_number, presence: true, format: { with: VALID_PHONE_REGEX }
```
```
validates :phone_number, presence: true, format: { with: VALID_PHONE_REGEX }
```
`presence: true`は削除しましょう。
with_options内に記載いただいているため不要な記述となります。

Kentozzz commented on spec/factories/order_addresses.rb:
```diff
@@ -0,0 +1,10 @@
+    post_code     { '020-0104' }
+    prefecture_id { 2 }
+    city          { '盛岡市' }
+    house_number  { '1-1-1' }
```
正確なテストを実施するために建物名も生成しましょう！

Kentozzz commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,87 @@
+        @order_address.post_code = '123ー123４'
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include('Post code is invalid. Include hyphen(-)')
+      end
```
追加で以下のテストをお願いします！
```
post_codeはハイフンが無ければ購入できない
```
※半角数値にハイフンを無くした状態でテストしましょう。

Kentozzz commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,87 @@
+    user = FactoryBot.create(:user)
+    # ダミーイメージを保存済みのitem情報を作り、item_idに渡す
+    item = FactoryBot.build(:item)
+    item.image = fixture_file_upload('public/images/test_image.png')
```
```
item.image = fixture_file_upload('public/images/test_image.png')
```
こちらの記述は不要かと存じますので削除しましょう！
itemのFactoryBotで画像がすでに生成されているためです。

Kentozzz reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

55Suape committed:
itemsコントローラー修正（同一条件式をまとめる）

## app/controllers/items_controller.rb
- before_action でuser_checkメソッドを追記し、edit・destroyアクションだけに共通のチェック(出品者とカレントユーザー情報の相違確認)を行う

55Suape committed:
order_addressモデル修正（不要な記述トルツメ）

## app/models/order_address.rb
with_options内に記載済みのため、バリデーションphone_numberのpresence: true,トルツメ

55Suape committed:
order_addressモデル単体テスト修正

## spec/factories/order_addresses.rb
- 建物名の値を生成

## spec/models/order_address_spec.rb
- 画像はitemのFactoryBotにて生成済みのためitem.image =・・・.png')の記述をトルツメ
- 「 post_codeはハイフンが無ければ購入できない」項目追記

55Suape commented:
# What
商品購入機能の実装

# Why
コードレビューにてコメントをありがとうございます。
下記修正を反映したため、ご確認をお願いします。

- [ ] _app/controllers/items_controller.rb_
before_action でuser_checkメソッドを追記し、edit・destroyアクションだけに共通のチェック(出品者とカレントユーザー情報の相違確認)を行う

- [ ] _aapp/models/order_address.rb_
with_options内に記載済みのため、バリデーションphone_numberのpresence: true,トルツメ

- [ ] _aspec/factories/order_addresses.rb_
建物名の値を生成

- [ ] _aspec/models/order_address_spec.rb_
画像はitemのFactoryBotにて生成済みのため、item.image =・・・.png')の記述をトルツメ
「 post_codeはハイフンが無ければ購入できない」項目追記

▼単体テスト結果
https://gyazo.com/5cd588024e1f7aaba56f82e8c12d56cf

Kentozzz reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
このブランチをマージし、masterブランチでプルを行いましょう！
機能が反映されていることを確認し、引き続き実装頑張っていきましょう！
